File: app/src/main/java/io/euphoria/xkcd/app/control/RoomController.java
package io.euphoria.xkcd.app.control;

import android.content.Context;
import android.os.Handler;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import io.euphoria.xkcd.app.connection.Connection;
import io.euphoria.xkcd.app.connection.ConnectionListener;
import io.euphoria.xkcd.app.connection.ConnectionManager;
import io.euphoria.xkcd.app.connection.ConnectionStatus;
import io.euphoria.xkcd.app.connection.event.CloseEvent;
import io.euphoria.xkcd.app.connection.event.IdentityEvent;
import io.euphoria.xkcd.app.connection.event.LogEvent;
import io.euphoria.xkcd.app.connection.event.MessageEvent;
import io.euphoria.xkcd.app.connection.event.NickChangeEvent;
import io.euphoria.xkcd.app.connection.event.OpenEvent;
import io.euphoria.xkcd.app.connection.event.PresenceChangeEvent;
import io.euphoria.xkcd.app.ui.RoomUI;
import io.euphoria.xkcd.app.ui.RoomUIManager;
import io.euphoria.xkcd.app.ui.UIListener;
import io.euphoria.xkcd.app.ui.event.LogRequestEvent;
import io.euphoria.xkcd.app.ui.event.MessageSendEvent;
import io.euphoria.xkcd.app.ui.event.NewNickEvent;
import io.euphoria.xkcd.app.ui.event.RoomSwitchEvent;
import io.euphoria.xkcd.app.ui.event.UICloseEvent;

/** Created by Xyzzy on 2017-03-19. */

public class RoomController {

    public static final int DEFAULT_LOG_REQUEST_AMOUNT = 50;

    private final Context context;
    private final Handler handler;
    private final RoomUIManager uiManager;
    private final ConnectionManager connManager;
    private final Set<String> openRooms;

    public RoomController(Context ctx, RoomUIManager uiManager, ConnectionManager connManager) {
        this.context = ctx;
        this.handler = new Handler(ctx.getMainLooper());
        this.uiManager = uiManager;
        this.connManager = connManager;
        this.openRooms = new HashSet<>();
    }

    public Context getContext() {
        return context;
    }

    public RoomUIManager getRoomUIManager() {
        return uiManager;
    }

    public ConnectionManager getConnectionManager() {
        return connManager;
    }

    public void openRoom(String roomName) {
        Connection conn = connManager.connect(roomName);
        RoomUI ui = uiManager.getRoomUI(roomName);
        if (openRooms.add(roomName)) link(conn, ui);
    }

    public void closeRoom(String roomName) {
        Connection conn = connManager.getConnection(roomName);
        if (conn != null) conn.close();
        RoomUI ui = uiManager.getRoomUI(roomName);
        if (ui != null) ui.close();
        openRooms.remove(roomName);
    }

    public void shutdown() {
        connManager.shutdown();
        List<String> rooms = new ArrayList<>(openRooms);
        for (String roomName : rooms) {
            closeRoom(roomName);
        }
    }

    public void invokeLater(Runnable cb) {
        handler.post(cb);
    }

    public void invokeLater(Runnable cb, long delay) {
        handler.postDelayed(cb, delay);
    }

    protected void link(final Connection conn, final RoomUI ui) {
        ui.setConnectionStatus(ConnectionStatus.CONNECTING);
        conn.addEventListener(new ConnectionListener() {
            @Override
            public void onOpen(OpenEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        ui.setConnectionStatus(ConnectionStatus.CONNECTED);
                    }
                });
            }

            @Override
            public void onIdentity(final IdentityEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        ui.setIdentity(evt.getIdentity());
                        ui.showNicks(Collections.singletonList(evt.getIdentity()));
                    }
                });
            }

            @Override
            public void onNickChange(final NickChangeEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        ui.showNicks(Collections.singletonList(evt.getSession()));
                    }
                });
            }

            @Override
            public void onMessage(final MessageEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        ui.showMessages(Collections.singletonList(evt.getMessage()));
                    }
                });
            }

            @Override
            public void onPresenceChange(final PresenceChangeEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        if (evt.isPresent()) {
                            ui.showNicks(evt.getSessions());
                        } else {
                            ui.removeNicks(evt.getSessions());
                        }
                    }
                });
            }

            @Override
            public void onLogEvent(final LogEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        ui.showMessages(evt.getMessages());
                    }
                });
            }

            @Override
            public void onClose(final CloseEvent evt) {
                invokeLater(new Runnable() {
                    @Override
                    public void run() {
                        ui.setConnectionStatus(evt.isFinal() ? ConnectionStatus.DISCONNECTED :
                                                               ConnectionStatus.RECONNECTING);
                    }
                });
            }
        });
        ui.addEventListener(new UIListener() {
            @Override
            public void onNewNick(NewNickEvent evt) {
                conn.setNick(evt.getNewNick());
            }

            @Override
            public void onMessageSend(MessageSendEvent evt) {
                conn.postMessage(evt.getText(), evt.getParent());
            }

            @Override
            public void onLogRequest(LogRequestEvent evt) {
                conn.requestLogs(evt.getBefore(), DEFAULT_LOG_REQUEST_AMOUNT);
            }

            @Override
            public void onRoomSwitch(RoomSwitchEvent evt) {
                openRoom(evt.getRoomName());
            }

            @Override
            public void onClose(UICloseEvent evt) {
                closeRoom(evt.getRoomUI().getRoomName());
            }
        });
    }

}

File: app/src/main/java/io/euphoria/xkcd/app/control/RoomController.kt
package io.euphoria.xkcd.app.control

import android.content.Context
import android.os.Handler
import io.euphoria.xkcd.app.connection.Connection
import io.euphoria.xkcd.app.connection.ConnectionListener
import io.euphoria.xkcd.app.connection.ConnectionManager
import io.euphoria.xkcd.app.connection.ConnectionStatus
import io.euphoria.xkcd.app.connection.event.*
import io.euphoria.xkcd.app.ui.RoomUI
import io.euphoria.xkcd.app.ui.RoomUIManager
import io.euphoria.xkcd.app.ui.UIListener
import io.euphoria.xkcd.app.ui.event.*
import java.util.*

/** Created by Xyzzy on 2017-03-19.  */
class RoomController constructor(
    val context: Context,
    uiManager: RoomUIManager,
    connManager: ConnectionManager
) {
    private val handler: Handler
    val roomUIManager: RoomUIManager
    val connectionManager: ConnectionManager
    private val openRooms: MutableSet<String>

    fun openRoom(roomName: String) {
        val conn = connectionManager.connect(roomName)
        val ui = roomUIManager.getRoomUI(roomName)
        if (openRooms.add(roomName)) link(conn, ui)
    }

    fun closeRoom(roomName: String) {
        connectionManager.getConnection(roomName)?.close()
        roomUIManager.getRoomUI(roomName).close()
        openRooms.remove(roomName)
    }

    fun shutdown() {
        connectionManager.shutdown()
        val rooms: List<String> =
            ArrayList(openRooms)
        for (roomName: String in rooms) {
            closeRoom(roomName)
        }
    }

    fun invokeLater(cb: Runnable) {
        handler.post(cb)
    }

    fun invokeLater(cb: Runnable, delay: Long) {
        handler.postDelayed(cb, delay)
    }

    protected fun link(
        conn: Connection,
        ui: RoomUI
    ) {
        ui.setConnectionStatus(ConnectionStatus.CONNECTING)
        conn.addEventListener(object : ConnectionListener {
            override fun onOpen(evt: OpenEvent?) {
                invokeLater(Runnable { ui.setConnectionStatus(ConnectionStatus.CONNECTED) })
            }

            override fun onIdentity(evt: IdentityEvent) {
                invokeLater(Runnable { ui.showNicks(listOf(evt.identity)) })
            }

            override fun onNickChange(evt: NickChangeEvent) {
                invokeLater(Runnable { ui.showNicks(listOf(evt.session)) })
            }

            override fun onMessage(evt: MessageEvent) {
                invokeLater(Runnable { ui.showMessages(listOf(evt.message)) })
            }

            override fun onPresenceChange(evt: PresenceChangeEvent) {
                invokeLater(Runnable {
                    if (evt.isPresent) {
                        ui.showNicks(evt.sessions)
                    } else {
                        ui.removeNicks(evt.sessions)
                    }
                })
            }

            override fun onLogEvent(evt: LogEvent) {
                invokeLater(Runnable { ui.showMessages(evt.messages) })
            }

            override fun onClose(evt: CloseEvent) {
                invokeLater(Runnable { ui.setConnectionStatus(if (evt.isFinal) ConnectionStatus.DISCONNECTED else ConnectionStatus.RECONNECTING) })
            }
        })
        ui.addEventListener(object : UIListener {
            override fun onNewNick(evt: NewNickEvent) {
                conn.setNick(evt.newNick)
            }

            override fun onMessageSend(evt: MessageSendEvent) {
                conn.postMessage(evt.text, evt.parent)
            }

            override fun onLogRequest(evt: LogRequestEvent) {
                conn.requestLogs(
                    evt.before,
                    DEFAULT_LOG_REQUEST_AMOUNT
                )
            }

            override fun onRoomSwitch(evt: RoomSwitchEvent) {
                openRoom(evt.roomName)
            }

            override fun onClose(evt: UICloseEvent) {
                evt.roomUI?.roomName?.let { closeRoom(it) }
            }
        })
    }

    companion object {
        const val DEFAULT_LOG_REQUEST_AMOUNT: Int = 50
    }

    init {
        handler = Handler(context.mainLooper)
        roomUIManager = uiManager
        connectionManager = connManager
        openRooms = HashSet()
    }
}
